azure-spring-cloud-maven-plugin/src/main/java/com/microsoft/azure/maven/springcloud/DeployMojo.java

```diff
@@ -89,65 +90,111 @@ protected void doExecute() throws MojoExecutionException, AzureExecutionExceptio
+        final SpringCloudDeployment.Updater deploymentModifier = (toCreateDeployment ? deployment.create() : deployment.update())
             .configEnvironmentVariables(env)
             .configJvmOptions(jvmOptions)
             .configScaleSettings(scaleSettings)
             .configRuntimeVersion(runtimeVersion)
-            .configAppArtifactPath(artifactPath)
-            .commit();
-
-        app.update()
+            .configArtifact(artifactUploader.getArtifact());
+        final SpringCloudApp.Updater appUpdater = app.update()
             .activate(Optional.ofNullable(appConfig.getActiveDeploymentName()).orElse(deploymentName))
             .setPublic(appConfig.isPublic())
-            .enablePersistentDisk(enableDisk)
-            .commit();
+            .enablePersistentDisk(enableDisk);
+
+        if (toCreateApp) {
+            tasks.add(new AzureTask<Void>(CREATE_APP_TITLE, () -> {
+                getLog().info(CREATE_APP_DOING);
+                appCreator.commit();
+                getLog().info(CREATE_APP_DONE);
+            }));
+        }
+        tasks.add(new AzureTask<Void>(UPLOAD_ARTIFACT_TITLE, () -> {
+            getLog().info(UPLOAD_ARTIFACT_DOING);
+            artifactUploader.commit();
+            getLog().info(UPLOAD_ARTIFACT_DONE);
+        }));
+        tasks.add(new AzureTask<Void>(toCreateDeployment ? CREATE_DEPLOYMENT_TITLE : UPDATE_DEPLOYMENT_TITLE, () -> {
+            getLog().info(toCreateDeployment ? CREATE_DEPLOYMENT_DOING : UPDATE_DEPLOYMENT_DOING);
+            deploymentModifier.commit();
+            getLog().info(toCreateDeployment ? CREATE_DEPLOYMENT_DONE : UPDATE_DEPLOYMENT_DONE);
+        }));
+        if (!appUpdater.isSkippable()) {
+            tasks.add(new AzureTask<Void>(UPDATE_APP_TITLE, () -> {
+                getLog().info(UPDATE_APP_DOING);
+                appUpdater.commit();
+                getLog().info(UPDATE_APP_DONE);
+                getLog().warn(UPDATE_APP_WARNING);
+            }));
+        }
+
+        tasks.add(new AzureTask<Void>(() -> {
+            if (!noWait) {
+                getLog().info(GET_DEPLOYMENT_STATUS_DOING);
+                if (!deployment.waitUntilReady(GET_STATUS_TIMEOUT)) {
+                    getLog().warn(GET_DEPLOYMENT_STATUS_TIMEOUT);
+                }
+            }
+            printStatus(deployment);
+            printPublicUrl(app);
+        }));
+        if (this.confirmDeploy(tasks)) {
+            tasks.forEach(AzureTask::go);
+        }
+    }
 
-        if (!noWait) {
-            deployment.waitUntilReady(GET_STATUS_TIMEOUT);
+    protected boolean confirmDeploy(List<AzureTask<?>> tasks) throws MojoFailureException {
+        try (IPrompter prompter = new DefaultPrompter()) {
+            getLog().info(CONFIRM_PROMPT_START);
+            tasks.stream().filter(t -> StringUtils.isNotEmpty(t.getTitle())).forEach((t) -> System.out.printf("\t*  %s%n", t.getTitle()));
```

Comment:
is System.out.printf() necessary ?

Reply:
yes.

Reply:
how about to replace it with getLog().info(...) ?